ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp

```diff
@@ -486,13 +496,24 @@ bool VirtualizationRoot_VnodeIsOnAllowedFilesystem(vnode_t vnode)
 
 static const char* GetRelativePath(const char* path, const char* root)
 {
-    assert(strlen(path) >= strlen(root));
+    size_t rootLength = strlen(root);
+    size_t pathLength = strlen(path);
+    if (pathLength < rootLength || 0 != memcmp(path, root, rootLength))
+    {
+        KextLog_Error("GetRelativePath: root path '%s' is not a prefix of path '%s'\n", root, path);
+        return nullptr;
+    }
     
-    const char* relativePath = path + strlen(root);
+    const char* relativePath = path + rootLength;
     if (relativePath[0] == '/')
     {
         relativePath++;
     }
+    else if (rootLength > 0 && root[rootLength - 1] != '/' && pathLength > rootLength)
```

Comment:
Can this check:

`else if (rootLength > 0 && root[rootLength - 1] != '/' && pathLength > rootLength)`

Be simplified to:

`else if (rootLength != pathLength)`

My thinking was that the root path can only be a parent of `path` at this point  if either:

- `relativePath[0] == '/'` is true
- `rootLength == pathLength` (i.e. `relativePath[0]` is `'\0'`)

Can you think of a scenario I'm missing?

Reply:
In principle, if root is `"/"`, and path is `"/something/else"`, `relativePath[0]` will not be `'/'` and the lengths won't match. Although of course at the moment that would mean a virtualisation root at /, which isn't terribly practical. I don't *think* `vn_getpath()` should be trailing slashes for any other kind of path, so assuming we'll never call `GetRelativePath()` with `root = "/"` we can make that change. I'm not sure it's worth doing that for performance's sake, considering `relativePath[0] == '/'` will be the hottest code path by far.

Reply:
> In principle, if root is "/", and path is "/something/else", relativePath[0] will not be '/' and the lengths won't match.

Ahh, I'd been assuming that `root` would never end with `"/"` (even in the case of the root directory).

Reply:
I've definitely seen `vn_getpath` return `"/"` for the root vnode. The empty path would be relative, not absolute, so I guess that's why.

Reply:
> I've definitely seen vn_getpath return "/" for the root vnode. The empty path would be relative, not absolute, so I guess that's why.

Ok cool, thanks again for the details!